Cython/Compiler/PyrexTypes.py

```diff
@@ -1203,6 +1205,95 @@ def same_as(self, other_type, compare_base=True):
                 (not compare_base or self.base.same_as(other_type.base)))
 
 
+class HPyType(PyrexType):
+    #
+    #  Class for HPy handle type.
+    #
+
+    name = "hpy"
+    is_hpy = 1
+    default_value = "HPy_NULL"
+    declaration_value = "HPy_NULL"
+    buffer_defaults = None
+    is_extern = False
+    is_subclassed = False
+    is_gc_simple = False
+    builtin_trashcan = False  # builtin type using trashcan
+    needs_refcounting = True
+
+    def __str__(self):
+        return "HPy handle"
+
+    def __repr__(self):
+        return "<HPy>"
+
+    def can_coerce_to_pyobject(self, env):
+        return True
+
+    def can_coerce_from_pyobject(self, env):
+        return True
+
+    def default_coerced_ctype(self):
+        """The default C type that this Python type coerces to, or None."""
+        return None
+
+    def assignable_from(self, src_type):
+        # except for pointers, conversion will be attempted
+        return src_type.is_hpy
+
+    def declaration_code(self, entity_code,
+                         for_display = 0, dll_linkage = None, pyrex = 0):
+        if pyrex or for_display:
+            base_code = "object"
+        else:
+            base_code = public_decl("HPy", dll_linkage)
+            entity_code = "%s" % entity_code
+        return self.base_declaration_code(base_code, entity_code)
+
+    def as_pyobject(self, cname):
+        if (not self.is_complete()) or self.is_extension_type:
+            return "(PyObject *)" + cname
+        else:
+            return cname
+
+    def py_type_name(self):
+        return "hpy"
+
+    def __lt__(self, other):
+        """
+        Make sure we sort highest, as instance checking on py_type_name
+        ('object') is always true
+        """
+        return False
+
+    def global_init_code(self, entry, code):
+        code.put_init_var_to_py_none(entry, nanny=False)
+
+    def check_for_null_code(self, cname):
+        return cname
+
+    def generate_dup(self, code, hpy_ctx_cname, cname):
+        code.putln("HPy_Dup(%s, %s);" % (hpy_ctx_cname, cname))
+
+    def generate_close(self, code, hpy_ctx_cname, cname, have_gil):
```

Comment:
I know HPy doesn't use `INCREF` and `DECREF` because it's designed to work with non-reference counted implementations, but I think Cython will end up using them in the same place as incref and decref? Would it make sense to keep the interface in Cython as `generate_incref`/`generate_decref` (to match the existing interface)? Or is there a distinction I'm missing?

Reply:
I'd rather have Cython generate these things the HPy way, and then map those back to CPython with C macros. IIUC, `Dup()` is just `(INCREF(ref), ref)` (for which we have `__Pyx_NewRef()`), `Close()` is just `DECREF()`. We should just adapt the refnanny macros for this and then use `new_ref = __Pyx_NewRef(old_ref)` wherever we currently use an assignment with `INCREF`. Even for CPython, that feels cleaner to me. We can also use an inline function for this to allow passing the results of non-trivial expressions directly (without requiring simple reference variables for macros).

Reply:
I think we can keep the `generate_incref/decref` interfaces.
In practice, inc/decref is very similar to dup/close. If inc/decref is used in a well-behaved way, it shouldn't be a problem.